src/com/ms/silverking/cloud/dht/daemon/storage/NamespaceStore.java

```diff
@@ -2683,892 +2893,1199 @@ private void moveToNext() {
+    return fileSegmentCompactor.getTrashSegments(nsDir);
+  }
+
+  @Override
+  public List<Integer> listCompactSegments() throws IOException {
+    return fileSegmentCompactor.getCompactSegments(nsDir);
+  }
+  ////////////////////////////////////
+  // SSNamespaceStore implementation
+
+  private static final int numPendingPutWorkers = Math.min(8, Runtime.getRuntime().availableProcessors());
+  private static final int maxWorkBatchSize = 128;
+  private static final BlockingQueue<PendingPut> pendingPuts = new ArrayBlockingQueue<>(
+      numPendingPutWorkers * maxWorkBatchSize);
+
+  static {
+    if (enablePendingPuts) {
+      Log.warningf("numPendingPutWorkers: %d", numPendingPutWorkers);
+      for (int i = 0; i < numPendingPutWorkers; i++) {
+        new PendingPutWorker(i, pendingPuts);
+      }
+    }
+  }
+
+  static class PendingPut {
+    final NamespaceStore nsStore;
+    final List<StorageValueAndParameters> values;
+    final KeyedOpResultListener resultListener;
+
+    PendingPut(NamespaceStore nsStore, List<StorageValueAndParameters> values, KeyedOpResultListener resultListener) {
+      this.nsStore = nsStore;
+      this.values = values;
+      this.resultListener = resultListener;
+    }
+  }
+
+  private List<StorageValueAndParameters> fixupCompression(List<StorageValueAndParameters> values) {
+    List<StorageValueAndParameters> _values;
+
+    _values = new ArrayList<>(values.size());
+    for (StorageValueAndParameters svp : values) {
+      StorageValueAndParameters _svp;
+      StorageValueAndParameters __svp;
+      int compressedLength;
+
+      if (svp.compressedSizeSet()) {
+        compressedLength = svp.getCompressedSize();
+      } else {
+        compressedLength = svp.getValue().remaining();
+      }
+      // The client disallows any compression usage when the compressed size equals or exceeds the
+      // uncompressed size. As ccss is per message, and this compression check is per value,
+      // we catch this case and modify the ccss for values where this applies.
+      if (compressedLength == svp.getUncompressedSize()) {
+        _svp = svp.ccss(CCSSUtil.updateCompression(svp.getCCSS(), Compression.NONE));
+      } else {
+        _svp = svp;
+      }
+      if (svpMapper != null && compressedLength > svpMapThreshold) {
+        try {
+          __svp = svpMapper.convertToMappedSVP(_svp);
+        } catch (Exception e) { // only IO thrown, but catch & tolerate anything
+          Log.logErrorWarning(e);
+          __svp = _svp;
+        }
+      } else {
+        __svp = _svp;
+      }
+      _values.add(__svp);
+    }
+    return _values;
+  }
+
+  private void addPendingPut(List<StorageValueAndParameters> values, KeyedOpResultListener resultListener) {
+    try {
+      //Log.warningf("Adding pending put %d", values.size());
+      pendingPuts.put(new PendingPut(this, fixupCompression(values), resultListener));
+    } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)